pandas/_libs/index.pyx

```diff
@@ -599,70 +600,91 @@ cpdef convert_scalar(ndarray arr, object value):
+    Consequently, labels are located by first locating each component on the
+    respective level, then locating for (the integer representation of) codes.
     """
-    def get_indexer(self, values):
-        # convert a MI to an ndarray
-        if hasattr(values, 'values'):
-            values = values.values
-        return super(MultiIndexObjectEngine, self).get_indexer(values)
-
-    cpdef get_loc(self, object val):
-
-        # convert a MI to an ndarray
-        if hasattr(val, 'values'):
-            val = val.values
-        return super(MultiIndexObjectEngine, self).get_loc(val)
+    def __init__(self, object levels, object labels,
+                 ndarray[uint64_t, ndim=1] offsets):
+        self.levels = levels
+        self.offsets = offsets
+
+        # Transform labels in a single array, and add 1 so that we are working
+        # with positive integers (-1 for NaN becomes 0):
+        codes = (np.array(labels, dtype='int64').T + 1).astype('uint64')
+        # Map each codes combination in the index to an integer unambiguously
+        # (no collisions possible), based on the "offsets", which describe the
+        # number of bits to switch labels for each level:
+        lab_ints = self.codes_to_ints(codes)
+
+        # Initialize underlying index
+        self._base.__init__(self, lambda: lab_ints, len(lab_ints))
+
+    def get_indexer(self, object target, object method=None,
+                    object limit=None):
+        level_codes = [self.levels[lev].get_indexer(codes, method=method) + 1
+                       for lev, codes in enumerate(zip(*target))]
+
+        keys_int = self.codes_to_ints(np.array(level_codes, dtype='uint64').T)
+
+        # All methods (exact, backfill, pad) directly map to the respective
+        # methods of the underlying (integers) index...
+        if method is not None:
+            # but underlying backfill and pad methods require index and keys
+            # to be sorted. The index already is (checked in
+            # Index._get_fill_indexer), so sort keys (integer representations):
+            order = np.argsort(keys_int)
+            keys_int = keys_int[order]
+            sup_meth = getattr(self._base, 'get_{}_indexer'.format(method))
+            indexer = sup_meth(self, keys_int, limit=limit)
+            indexer = indexer[order]
+        else:
+            indexer = self._base.get_indexer(self, keys_int)
 
+        return indexer
 
-cdef class MultiIndexHashEngine(ObjectEngine):
-    """
-    Use a hashing based MultiIndex impl
-    but use the IndexEngine for computation
+    def get_loc(self, object key):
+        if is_definitely_invalid_key(key):
+            raise TypeError("'{key}' is an invalid key".format(key=key))
+        if not isinstance(key, tuple):
+            raise KeyError(key)
+        try:
+            indices = [0 if checknull(v) else self.levels[l].get_loc(v) + 1
+                       for l, v in enumerate(key)]
```

Comment:
You can iterate over the levels instead of indexing into them (it appears that getitem of Frozenlist is rather expensive):

```
In [23]: %timeit [(v, mi_small.levels[l]) for l, v in enumerate(key)]
3.58 µs ± 55.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [24]: %timeit [(v, l) for l, v in zip(mi_small.levels, key)]
922 ns ± 13.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
Another potential improvement ( but didn't time this) is to only do the `checknull` if needed (so if `get_loc` on the level raises an error, only then do the checknull check). This will improve a little bit the time for the case where there are no NaNs, but decrease for the NaN case (but i think this is more rare)